### DISCUSSION TITLE
Are my API keys safe with my setup?

### DISCUSSION THREAD

**Discussion by alioshr** [vercel/next.js#20095] [Help] [answered]
Hello there.

I have a simple mailer API set with ```nodemailer```. 

I read the docs regarding the best practices regarding where to place APIs, but I am still not sure about some details.

This is my setup:

I Have a handler with ```axios``` on the client side in a ```Components folder```:

On ```pages/api``` I have the following files:

```contact.js``` - > holding the ```nodemailer setup```
```key.json``` -> holding my ```API keys```

I import ```key.json``` inside ```contact.js``` to get my ```API keys``` in order to use ```nodemailer```.

Is this a safe way of placing my ```API keys```?

In case you guys want more info I can paste my code here.

Thanks in advance!

**Comment by jamesmosier** [comment, marked as answer]
That would be safe as no code within API routes is exposed to the browser.

An even better way would be to use [environment variables](https://nextjs.org/docs/basic-features/environment-variables) to expose secrets to server and client code.

**Reply by alioshr** [reply]
Thanks for you time here with me!

In my case there is no need to expose keys on the client, as my code runs just to get the email & name of the users through a ```form``` in one of my pages. So, on the client, I only collect the ```input``` values.

My API Keys are just there to config ```nodemailer```. 

Can/Should I config the enviromental variables to be used on ```nodemailer``` in this case?

**Reply by jamesmosier** [reply]
Yes i'd recommend it.

**Reply by alioshr** [reply]
Thanks!

One option is a setup on ```next.config.js```, adding an ```env``` configuration on my module.exports. 

```
module.exports = {
    env: {
        mailer: {
            key: "some key"
        }
    }
}
```

I liked this setup as I can separate stuff in objects, placing my API keys for nodemailer inside an object instance.
But I am not sure how could I separate public from non-public data on module.exports.

So I guess that I would take the most recent approach, setting .env files.

How can I define all these env variables in a way that I do not need to keep writting long var names, in order to make them unique? I would like to use this most recent approach, but somehow wrapping these variables in objects.

Another issue is that when I create a ```.env.local.nodemailer```, Vscode did not recognize the file and it did work... 

Do you have any workarounds for the naming extension? 

I have tried configuring my config file: 

```
"files.associations": {
            "*.js": "javascriptreact",
            "*.env.development": "env",
            "*.env.production": "env",
            "*.env.local": "env",
    },
```

**Reply by waptik** [reply]
> Thanks!
> 
> One option is a setup on `next.config.js`, adding an `env` configuration on my module.exports.
> 
> ```
> module.exports = {
>     env: {
>         mailer: {
>             key: "some key"
>         }
>     }
> }
> ```
> 
> I liked this setup as I can separate stuff in objects, placing my API keys for nodemailer inside an object instance.
> But I am not sure how could I separate public from non-public data on module.exports.
> 
> So I guess that I would take the most recent approach, setting .env files.
> 
> How can I define all these env variables in a way that I do not need to keep writting long var names, in order to make them unique? I would like to use this most recent approach, but somehow wrapping these variables in objects.
> 
> Another issue is that when I create a `.env.local.nodemailer`, Vscode did not recognize the file and it did work...
> 
> Do you have any workarounds for the naming extension?
> 
> I have tried configuring my config file:
> 
> ```
> "files.associations": {
>             "*.js": "javascriptreact",
>             "*.env.development": "env",
>             "*.env.production": "env",
>             "*.env.local": "env",
>     },
> ```

regarding naming extensions, i'd recommend you use [dotenv vscode extension](https://marketplace.visualstudio.com/items?itemName=mikestead.dotenv) it solves that issue for you.
Also, in local development, `.env.local` is more than enough.
